src/Common/StackTrace.cpp

```diff
@@ -244,102 +304,79 @@ size_t StackTrace::getOffset() const
 {
     if (size == 0)
         return callback("<Empty trace>");
 
-#if defined(__ELF__) && !defined(__FreeBSD__)
-    const DB::SymbolIndex & symbol_index = DB::SymbolIndex::instance();
-    std::unordered_map<std::string, DB::Dwarf> dwarfs;
-
     std::stringstream out;
 
     for (size_t i = offset; i < size; ++i)
     {
-        const void * virtual_addr = frames[i];
-        const auto * object = symbol_index.findObject(virtual_addr);
-        uintptr_t virtual_offset = object ? uintptr_t(object->address_begin) : 0;
-        const void * physical_addr = reinterpret_cast<const void *>(uintptr_t(virtual_addr) - virtual_offset);
-
+        const StackTrace::Frame & current_frame = frames[i];
         out << i << ". ";
 
-        if (object)
+        if (current_frame.file.has_value() && current_frame.line.has_value())
         {
-            if (std::filesystem::exists(object->name))
-            {
-                auto dwarf_it = dwarfs.try_emplace(object->name, *object->elf).first;
-
-                DB::Dwarf::LocationInfo location;
-                if (dwarf_it->second.findAddress(uintptr_t(physical_addr), location, DB::Dwarf::LocationInfoMode::FAST))
-                    out << location.file.toString() << ":" << location.line << ": ";
-            }
+            out << current_frame.file.value() << ":" << current_frame.line.value() << ": ";
         }
 
-        const auto * symbol = symbol_index.findSymbol(virtual_addr);
-        if (symbol)
+        if (current_frame.symbol.has_value())
         {
-            int status = 0;
-            out << demangle(symbol->name, status);
+            out << current_frame.symbol.value();
         }
-        else
-            out << "?";
-
-        out << " @ " << physical_addr;
-        out << " in " << (object ? object->name : "?");
 
-        callback(out.str());
-        out.str({});
-    }
-#else
-    std::stringstream out;
-
-    for (size_t i = offset; i < size; ++i)
-    {
-        const void * addr = frames[i];
-        out << i << ". " << addr;
+        out << " @ " << current_frame.physical_addr;
+        if (current_frame.object.has_value())
+        {
+            out << " in " << current_frame.object.value();
+        }
 
         callback(out.str());
         out.str({});
     }
-#endif
 }
 
-static std::string toStringImpl(const StackTrace::Frames & frames, size_t offset, size_t size)
+static std::string toStringImpl(const void * const * frame_pointers, size_t offset, size_t size)
 {
     std::stringstream out;
+    StackTrace::Frames frames;
+    StackTrace::symbolize(frame_pointers, offset, size, frames);
     toStringEveryLineImpl(frames, offset, size, [&](const std::string & str) { out << str << '\n'; });
     return out.str();
 }
 
 void StackTrace::toStringEveryLine(std::function<void(const std::string &)> callback) const
 {
+    Frames frames;
+    StackTrace::symbolize(frame_pointers.data(), offset, size, frames);
```

Comment:
It may make things worse...

DWARF::findAddress is very computationally heavy method, that's why it's better to print every next line as soon as it was calculated for that line.

Reply:
> It may make things worse...
> 
> DWARF::findAddress is very computationally heavy method, that's why it's better to print every next line as soon as it was calculated for that line.

@alexey-milovidov those values are usually small, thus keeping them around for a short period of time shouldn't be noticeable compared to this computationally heavy method.

Reply:
I mean - it will prevent logging every frame as soon as it's symbolized.

Reply:
@alexey-milovidov why that's important?

Reply:
Otherwise there is a chance that the program will fail while printing nth stack frame and we won't see 1st..n-1th frames. This has happened in our practice.

Reply:
@alexey-milovidov ok, returned the previous implementation that did that